gobblin-temporal/src/main/java/org/apache/gobblin/temporal/yarn/DynamicScalingYarnService.java

```diff
@@ -62,6 +82,73 @@ protected synchronized void requestInitialContainers() {
+   *   <li> The container gets preempted by the ResourceManager </li>
+   *   <li> The container gets killed due to some reason, for example, if it runs over the allowed amount of virtual or physical memory </li>
+   * </ol>
+   * A replacement container is needed in all except the first case.
+   * </p>
+   */
+  @Override
+  protected void handleContainerCompletion(ContainerStatus containerStatus) {
+    ContainerId completedContainerId = containerStatus.getContainerId();
+    ContainerInfo completedContainerInfo = this.containerMap.remove(completedContainerId);
+
+    // Because callbacks are processed asynchronously, we might encounter situations where handleContainerCompletion()
+    // is called before onContainersAllocated(), resulting in the containerId missing from the containersMap.
+    // We use removedContainerIds to remember these containers and remove them from containerMap later
+    // when we call reviseWorkforcePlanAndRequestNewContainers method
+    if (completedContainerInfo == null) {
+      log.warn("Container {} not found in containerMap. This container onContainersCompleted() likely called before onContainersAllocated()",
+          completedContainerId);
+      this.removedContainerIds.add(completedContainerId);
+      return;
+    }
+
+    log.info("Container {} running profile {} completed with exit status {}",
+        completedContainerId, completedContainerInfo.getWorkerProfileName(), containerStatus.getExitStatus());
+
+    if (StringUtils.isNotBlank(containerStatus.getDiagnostics())) {
+      log.info("Container {} running profile {} completed with diagnostics: {}",
+          completedContainerId, completedContainerInfo.getWorkerProfileName(), containerStatus.getDiagnostics());
+    }
+
+    if (this.shutdownInProgress) {
+      log.info("Ignoring container completion for container {} as shutdown is in progress", completedContainerId);
+      return;
+    }
+
+    WorkerProfile workerProfile = completedContainerInfo.getWorkerProfile();
+
+    switch (containerStatus.getExitStatus()) {
+      case(ContainerExitStatus.ABORTED):
+        handleAbortedContainer(completedContainerId, completedContainerInfo);
+        break;
+      case(ContainerExitStatus.PREEMPTED):
+        log.info("Container {} for profile {} preempted, starting to launching a replacement container",
+            completedContainerId, completedContainerInfo.getWorkerProfileName());
+        requestContainersForWorkerProfile(workerProfile, 1);
+        break;
+      case(GENERAL_OOM_EXIT_STATUS_CODE):
+      case(ContainerExitStatus.KILLED_EXCEEDED_VMEM):
+      case(ContainerExitStatus.KILLED_EXCEEDED_PMEM):
+        handleContainerExitedWithOOM(completedContainerId, completedContainerInfo);
+        break;
+      case(1): // Same as linux exit status 1 Often occurs when launch_container.sh failed
+        log.info("Exit status 1.CompletedContainerInfo = {}", completedContainerInfo);
+        break;
+      default:
+        break;
```

Comment:
let's add a no-op for `KILLED_AFTER_APP_COMPLETION` & `SUCCESS` and add a log statement for default case, since there are other statuses like `DISKS_FAILED`, `KILLED_BY_CONTAINER_SCHEDULER` also in ContainerExitStatus, so having a log would be useful in case that also needs to be handled in future

```
  case ContainerExitStatus.KILLED_AFTER_APP_COMPLETION:
  case ContainerExitStatus.SUCCESS:
    break;

  default:
    // log any other unhandled completion code
    log.warn("Container {} exited with unhandled status code {}. ContainerInfo: {}",
        completedContainerId, containerStatus.getExitStatus(), completedContainerInfo);
    break;
```